csharp/src/SeedLang/X/VisitorHelper.cs

```diff
@@ -22,101 +23,138 @@
-    }
-
-    // Builds an eval statement from the eval token and the expression context.
-    internal static EvalStatement BuildEval(IToken evalToken, ParserRuleContext exprContext,
-                                            AbstractParseTreeVisitor<AstNode> visitor) {
-      var evalRange = RangeOfToken(evalToken);
-      var expr = visitor.Visit(exprContext) as Expression;
-      Debug.Assert(expr.Range is TextRange);
-      return Statement.Eval(expr, CombineRanges(evalRange, expr.Range as TextRange));
+    public VisitorHelper(IList<SyntaxToken> tokens) {
+      _syntaxTokens = tokens;
     }
 
-    // Sets grouping range for sub-expression to use. Only keeps the largest grouping range.
-    internal void SetGroupingRange(IToken leftParen, IToken rightParen) {
-      if (_groupingRange is null) {
-        _groupingRange = CombineRanges(RangeOfToken(leftParen), RangeOfToken(rightParen));
+    // Checks the child count of parser rule context. Throw an exception if the criterion is not
+    // met.
+    internal static void EnsureChildCountOfContext(ParserRuleContext context, int count) {
+      if (context.ChildCount != count) {
+        throw new ParseException($"Parse {context.GetType().Name} error.");
       }
     }
 
     // Builds a binary expression from the binary operator and expression contexts.
     //
     // The exprContexts parameter must contain exact 2 items: the left and right ExprContext.
-    internal BinaryExpression BuildBinary(BinaryOperator op, ParserRuleContext[] exprContexts,
+    internal BinaryExpression BuildBinary(IToken opToken, BinaryOperator op,
+                                          ParserRuleContext[] exprContexts,
                                           AbstractParseTreeVisitor<AstNode> visitor) {
       TextRange range = _groupingRange;
       _groupingRange = null;
 
       Debug.Assert(exprContexts.Length == 2);
-      var left = visitor.Visit(exprContexts[0]) as Expression;
-      var right = visitor.Visit(exprContexts[1]) as Expression;
-      if (range is null) {
-        Debug.Assert(left.Range is TextRange);
-        Debug.Assert(right.Range is TextRange);
-        range = CombineRanges(left.Range as TextRange, right.Range as TextRange);
+      var left = visitor.Visit(exprContexts[0]);
+      AddSyntaxToken(SyntaxType.Operator, CodeReferenceUtils.RangeOfToken(opToken));
+      var right = visitor.Visit(exprContexts[1]);
+
+      if (left is Expression leftExpr && right is Expression rightExpr) {
+        if (range is null) {
+          Debug.Assert(left.Range is TextRange);
+          Debug.Assert(right.Range is TextRange);
+          range = CodeReferenceUtils.CombineRanges(left.Range as TextRange,
+                                                   right.Range as TextRange);
+        }
+        return Expression.Binary(leftExpr, op, rightExpr, range);
       }
-      return Expression.Binary(left, op, right, range);
+      throw new ParseException("Invalid binary expression.");
```

Comment:
What are the exact causes that trigger this ParseException? If there are real cases, 

* The error message should be specific and should include the Range info in the source code...
* The error message should be something defined at SeedLang.Common.Message.cs
* The error message should be passed out to the client via Diagnostic info.

If not, why do we need a string message here? What's the message for?

Reply:
Same as above.